api/metadata_cache.go

```diff
@@ -17,52 +18,79 @@ import (
 const metadataCacheKey = "WPT-METADATA"
 
 type webappMetadataFetcher struct {
-	ctx    context.Context
-	client *http.Client
-	url    string
+	ctx        context.Context
+	client     *http.Client
+	url        string
+	gitHubUtil shared.GitHubUtil
 }
 
-func (f webappMetadataFetcher) Fetch() (res map[string][]byte, err error) {
-	metadataMap, err := getMetadataFromMemcache(f.ctx, f.client, f.url)
-	if err == nil && metadataMap != nil {
-		return metadataMap, nil
+func (f webappMetadataFetcher) Fetch() (sha *string, res map[string][]byte, err error) {
+	sha, metadataMap, err := getMetadataFromMemcache(f.ctx, f.client, f.url, f.gitHubUtil)
+	if err == nil && metadataMap != nil && sha != nil {
+		return sha, metadataMap, nil
 	}
 
-	return shared.CollectMetadataWithURL(f.client, f.url)
+	shaKey, err := f.gitHubUtil.GetWPTMetadataMasterSHA()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// CollectMetadataWithURL retrieves the content of the wpt-metadata repo from master by default.
+	res, err = shared.CollectMetadataWithURL(f.client, f.url)
```

Comment:
In addition, I know this was already here before, but would we ever hit this code? `getMetadataFromMemcache` also calls `CollectMetadataWithURL` if cache is missed, so it seems to me we have two duplicate fallback paths.

Reply:
Using `https://developer.github.com/v3/repos/contents/#get-archive-link` was also my origin consideration, however there are a slight complications here:

1. IIUC, this API only takes ref, which is a branch name. Even if the race condition happens, we have no way to refer to the `master` we get from `GetWPTMetadataMasterSHA` because `master` has changed. Also there is no way to get an archive from a `sha` for this [reason](https://stackoverflow.com/questions/26135216/why-isnt-there-a-git-clone-specific-commit-option/26135822#26135822)

2.Retrieving metadata this way kind of circles back to the origin issue with testing in searchcache. In metadata [search atoms](https://github.com/web-platform-tests/wpt.fyi/blob/master/api/query/atoms.go#L241), there is no way to pass in a mock object. We certainly don't have to make `webapp` and `searchcahe` work the same way (leave searchcache as it is), but I also don't like the inconsistency.

3. The race condition is also very unlikely to happen. These functions are almost called side-by-side. But of course, I would rather not have it.

Reply:
For your second comment, I think if it hits any of the `err != nil` case, such as [this one](https://github.com/web-platform-tests/wpt.fyi/pull/1927/files/74efe85420669ae0721eea02829e8ade45f6e762#diff-2813b30cfdffd1a9bf51da1d3bca7e06R49), this "second" fall back will hit. I would guess, the cache hits but the data is corrupted? So it's not a miss but `Unmarshal` would fail

Reply:
A SHA is also a valid ref. This seems to work: https://api.github.com/repos/web-platform-tests/wpt-metadata/tarball/69157cf699ab7a279b3c7a3a546ed9674c290996

Reply:
> 

For your second comment, I think if it hits any of the err != nil case, such as this one, this "second" fall back will hit. I would guess, the cache hits but the data is corrupted? So it's not a miss but Unmarshal would fail

Good point, but I think these two fallbacks should and can be consolidated.

Reply:
I actually spent lots of time googling ref vs SHA but should have actually just tried it. Thanks for it! Addressed both of your comments.